File: src/partition.py
"""
Algorithm 1: Diagonal Block Column Boundary Expansion
======================================================

Paper reference: Section III.B, Algorithm 1, Fig. 3.

Given a row-wise distributed CSR matrix, this module determines for each
MPI process the *diagonal block* column range [left, right) that balances:

  - **Communication**: columns outside [left, right) require remote vector
    elements (MPI messages).
  - **Computation**: each row should have at least ``lower_bound`` nonzeros
    inside the diagonal block so that local work is balanced.

The algorithm proceeds in four steps:

  S1. Count diag / off-diag nonzeros per local row using initial boundaries
      [r_start, r_end).
  S2. Compute ``lower_bound = Allreduce_max(max_i diag_nnz[i])``, capped at N.
  S3. Greedily expand the left / right boundary one column at a time,
      always choosing the side that benefits more *deficient* rows.
  S4. Return (left, right) for this rank.

After Algorithm 1, every process knows which columns it "owns".  Algorithm 2
then builds the communication schedule.
"""

from __future__ import annotations

import logging
from typing import Tuple

import numpy as np

logger = logging.getLogger(__name__)


def diagonal_block_expand(
    rowptr: np.ndarray,       # int32, shape (nlocal+1,)
    colidx: np.ndarray,       # int32, shape (nnz_local,)
    r_start: int,             # first global row owned by this process
    r_end: int,               # one past last global row
    ncols: int,               # total columns in the global matrix (N)
    comm,                     # MPI intra-node communicator
) -> Tuple[int, int]:
    """Run Algorithm 1 and return (left_boundary, right_boundary).

    Parameters
    ----------
    rowptr : np.ndarray
        CSR row pointers for locally-owned rows, length = nlocal + 1.
    colidx : np.ndarray
        CSR column indices for locally-owned rows.
    r_start : int
        First global row index owned by this rank (inclusive).
    r_end : int
        One past the last global row index (exclusive).
    ncols : int
        Number of columns in the global matrix.
    comm :
        MPI communicator (e.g. MPI.COMM_WORLD).

    Returns
    -------
    (left, right) : Tuple[int, int]
        Column range [left, right) that forms the diagonal block for this
        process.  Columns in this range are *locally owned*; all other
        columns require communication.
    """
    nlocal = r_end - r_start
    if nlocal == 0:
        return 0, 0

    # ── S1: Count diag / off-diag per row (initial boundaries) ──
    left = r_start
    right = min(r_end, ncols)

    total_nnz_per_row = np.diff(rowptr).astype(np.int32)  # length nlocal

    diag, off_left, off_right = _count_blocks(
        rowptr, colidx, nlocal, left, right,
    )

    # ── S2: lower_bound = max(max_i diag_nnz[i]), reduced across ranks ──
    local_max = int(diag.max()) if nlocal > 0 else 0
    lower_bound = comm.allreduce(local_max, op=_mpi_max_op(comm))
    lower_bound = min(lower_bound, ncols)
    logger.info(
        "Algorithm 1 init: left=%d right=%d local_max_diag=%d lower_bound=%d",
        left, right, local_max, lower_bound,
    )

    # ── S3: Greedy boundary expansion ──
    # Pre-build column→rows index so _update_diag_left/right and
    # _count_rows_with_col are O(rows_with_col) instead of O(nlocal).
    col_to_rows = _build_col_to_rows(rowptr, colidx, nlocal)

    max_iter = ncols  # safety cap
    for iteration in range(max_iter):
        # Identify rows still below the lower bound.
        # Cap the per-row target at total_nnz_per_row[i] —
        # a row cannot gain more diag nonzeros than it has nonzeros.
        target = np.minimum(lower_bound, total_nnz_per_row)
        deficit = np.maximum(target - diag, 0)
        deficient_rows = np.where(deficit > 0)[0]

        if len(deficient_rows) == 0:
            logger.info("Algorithm 1 converged after %d iterations", iteration)
            break

        deficient_set = set(deficient_rows.tolist())

        # ── Count how many deficient rows would benefit from expanding
        #     left by one column or right by one column ──
        left_gain = 0
        if left > 0:
            left_gain = _count_rows_with_col(
                col_to_rows, deficient_set, left - 1,
            )
        right_gain = 0
        if right < ncols:
            right_gain = _count_rows_with_col(
                col_to_rows, deficient_set, right,
            )

        if left_gain == 0 and right_gain == 0:
            logger.info(
                "Algorithm 1 stalled at iter %d: %d rows still deficient",
                iteration, len(deficient_rows),
            )
            break

        # Expand in the direction that helps more deficient rows
        if left_gain >= right_gain and left > 0:
            left -= 1
            _update_diag_left(col_to_rows, left, diag, off_left)
        elif right < ncols:
            _update_diag_right(col_to_rows, right, diag, off_right)
            right += 1
        else:
            break  # cannot expand either direction

    else:
        logger.warning("Algorithm 1 hit iteration limit (%d)", max_iter)

    logger.info(
        "Algorithm 1 result: left=%d right=%d  (initial=[%d,%d))",
        left, right, r_start, min(r_end, ncols),
    )
    return left, right


# ──────────────────────────────────────────────────────────────────────
# Internal helpers
# ──────────────────────────────────────────────────────────────────────

def _count_blocks(
    rowptr: np.ndarray, colidx: np.ndarray,
    nlocal: int, left: int, right: int,
) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Count diag, off_left, off_right nonzeros per local row.

    - diag[i]:     nonzeros with left <= col < right
    - off_left[i]: nonzeros with col < left
    - off_right[i]:nonzeros with col >= right
    """
    diag = np.zeros(nlocal, dtype=np.int32)
    off_left = np.zeros(nlocal, dtype=np.int32)
    off_right = np.zeros(nlocal, dtype=np.int32)

    for i in range(nlocal):
        for j in range(rowptr[i], rowptr[i + 1]):
            c = colidx[j]
            if c < left:
                off_left[i] += 1
            elif c >= right:
                off_right[i] += 1
            else:
                diag[i] += 1

    return diag, off_left, off_right


def _build_col_to_rows(
    rowptr: np.ndarray, colidx: np.ndarray,
    nlocal: int,
) -> dict:
    """Build a mapping from column index to list of local rows containing it.

    Returns a dict {col: np.ndarray of local row indices}.
    Only columns that appear in the local partition get an entry.
    """
    from collections import defaultdict

    col_buckets: defaultdict[int, list[int]] = defaultdict(list)
    for i in range(nlocal):
        for j in range(rowptr[i], rowptr[i + 1]):
            col_buckets[colidx[j]].append(i)

    return {c: np.array(rows, dtype=np.int32) for c, rows in col_buckets.items()}


def _count_rows_with_col(
    col_to_rows: dict,
    deficient_set: set,
    col: int,
) -> int:
    """Count how many deficient rows have a nonzero in column *col*."""
    row_list = col_to_rows.get(col)
    if row_list is None:
        return 0
    return sum(1 for r in row_list if r in deficient_set)


def _update_diag_left(
    col_to_rows: dict,
    col: int,
    diag: np.ndarray,
    off_left: np.ndarray,
) -> None:
    """Column *col* just moved from off_left into diag.  Update counters."""
    row_list = col_to_rows.get(col)
    if row_list is not None:
        for i in row_list:
            off_left[i] -= 1
            diag[i] += 1


def _update_diag_right(
    col_to_rows: dict,
    col: int,
    diag: np.ndarray,
    off_right: np.ndarray,
) -> None:
    """Column *col* just moved from off_right into diag.  Update counters."""
    row_list = col_to_rows.get(col)
    if row_list is not None:
        for i in row_list:
            off_right[i] -= 1
            diag[i] += 1


def _mpi_max_op(comm):
    """Return MPI.MAX for the given communicator.

    If MPI is not available (e.g. in unit-test mocks), returns None;
    the mock's allreduce should ignore the op parameter.
    """
    try:
        from mpi4py import MPI
        return MPI.MAX
    except (ImportError, RuntimeError):
        return None  # for mock comms in unit tests

File: src/partition.cpp
#include "partition.hpp"

#include <algorithm>
#include <cstdio>
#include <unordered_map>
#include <unordered_set>
#include <vector>

namespace distspmv {

std::pair<idx_t, idx_t> diagonal_block_expand(
    const idx_t* rowptr, const idx_t* colidx,
    idx_t nlocal, idx_t r_start, idx_t r_end, idx_t ncols,
    MPI_Comm comm) {

    if (nlocal == 0) return {0, 0};

    // ── S1: Initialize boundaries & count per-row ──
    idx_t left = r_start;
    idx_t right = std::min(r_end, ncols);

    std::vector<idx_t> diag(nlocal, 0);
    std::vector<idx_t> off_left(nlocal, 0);
    std::vector<idx_t> off_right(nlocal, 0);
    std::vector<idx_t> total_nnz(nlocal, 0);

    for (idx_t i = 0; i < nlocal; ++i) {
        total_nnz[i] = rowptr[i + 1] - rowptr[i];
        for (idx_t j = rowptr[i]; j < rowptr[i + 1]; ++j) {
            idx_t c = colidx[j];
            if (c < left)       off_left[i]++;
            else if (c >= right) off_right[i]++;
            else                 diag[i]++;
        }
    }

    // ── S2: Compute lower_bound ──
    idx_t local_max = 0;
    for (idx_t i = 0; i < nlocal; ++i) {
        if (diag[i] > local_max) local_max = diag[i];
    }

    idx_t lower_bound = 0;
    MPI_Allreduce(&local_max, &lower_bound, 1, MPI_INT, MPI_MAX, comm);
    if (lower_bound > ncols) lower_bound = ncols;

    int rank;
    MPI_Comm_rank(comm, &rank);
    if (rank == 0) {
        std::printf("[rank %2d] Algo1 init: left=%d right=%d local_max_diag=%d "
                    "lower_bound=%d\n",
                    rank, left, right, local_max, lower_bound);
    }

    // ── S3: Build column-to-rows index for fast lookup ──
    // Only for columns that appear in this local partition
    std::unordered_map<idx_t, std::vector<idx_t>> col_to_rows;
    for (idx_t i = 0; i < nlocal; ++i) {
        for (idx_t j = rowptr[i]; j < rowptr[i + 1]; ++j) {
            idx_t c = colidx[j];
            col_to_rows[c].push_back(i);
        }
    }

    // ── S4: Greedy boundary expansion ──
    idx_t max_iter = ncols;
    for (idx_t iter = 0; iter < max_iter; ++iter) {
        // Find deficient rows
        std::unordered_set<idx_t> deficient;
        std::vector<idx_t> target(nlocal);
        for (idx_t i = 0; i < nlocal; ++i) {
            target[i] = std::min(lower_bound, total_nnz[i]);
            if (diag[i] < target[i]) {
                deficient.insert(i);
            }
        }

        if (deficient.empty()) {
            if (rank == 0) {
                std::printf("[rank %2d] Algo1 converged after %d iterations\n",
                            rank, iter);
            }
            break;
        }

        // Count gains
        idx_t left_gain = 0;
        if (left > 0) {
            auto it = col_to_rows.find(left - 1);
            if (it != col_to_rows.end()) {
                for (idx_t r : it->second) {
                    if (deficient.count(r)) left_gain++;
                }
            }
        }

        idx_t right_gain = 0;
        if (right < ncols) {
            auto it = col_to_rows.find(right);
            if (it != col_to_rows.end()) {
                for (idx_t r : it->second) {
                    if (deficient.count(r)) right_gain++;
                }
            }
        }

        if (left_gain == 0 && right_gain == 0) {
            if (rank == 0) {
                std::printf("[rank %2d] Algo1 stalled at iter %d: %zu rows "
                            "deficient\n",
                            rank, iter, deficient.size());
            }
            break;
        }

        // Expand toward larger gain
        if (left_gain >= right_gain && left > 0) {
            left--;
            auto it = col_to_rows.find(left);
            if (it != col_to_rows.end()) {
                for (idx_t r : it->second) {
                    off_left[r]--;
                    diag[r]++;
                }
            }
        } else if (right < ncols) {
            auto it = col_to_rows.find(right);
            if (it != col_to_rows.end()) {
                for (idx_t r : it->second) {
                    off_right[r]--;
                    diag[r]++;
                }
            }
            right++;
        } else {
            break;
        }
    }

    if (rank == 0) {
        std::printf("[rank %2d] Algo1 result: left=%d right=%d "
                    "(initial [%d,%d))\n",
                    rank, left, right, r_start, std::min(r_end, ncols));
    }

    return {left, right};
}

}  // namespace distspmv
